### ISSUE TITLE
Any way to get connectionString/endpoint/labelFilter dynamically with mode="Expand"?

### ISSUE THREAD

**Issue by yar-shukan** [aspnet/MicrosoftConfigurationBuilders#124]
Hi @zhenlan  and @StephenMolloy ,

We're forced to use **`mode="Expand"`** for several reasons (among which is WCF related sections). Our `configBuilders` looks like this:
```
 <configBuilders>
    <builders>
      <add name="AzureAppConfiguration"
           endpoint="__AppConfigurationUrl__"
           connectionString="__AppConfigurationConnectionString__"
           labelFilter="__Environment__"
           mode="Expand"
           type="Microsoft.Configuration.ConfigurationBuilders.AzureAppConfigurationBuilder, Microsoft.Configuration.ConfigurationBuilders.AzureAppConfiguration, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
    </builders>
  </configBuilders>
```

**Every time we need to run/debug app locally we have to modify Web.config/App.config to replace tokens with actual values. This is very inconvenient, error prone and sometimes people can commit these lines. This becomes a real issue when you have dozens of repos (microservices) and you need to do that everywhere.**

Is there any way to handle/workaround this design issue and get endpoint/connectionString/labelFilter dynamically e.g. from environment variables when still using **mode="Expand"**? I know that [documentation says](https://github.com/aspnet/MicrosoftConfigurationBuilders#appsettings-parameters): 

> If a builder is in Expand mode and processing the appSettings section, then this feature is disabled.

But still... that sounds like a design gap/issue to me. Any way to fix/workaround this experience?

**Comment by StephenMolloy** [comment]
**TLDR;** Your config builder is processing a WCF section, not appSettings. There shouldn't be a problem here.


The design is what it is. The reason for that is because we are trying to shoe-horn this cool new dynamic stuff into a configuration system that was designed in a different world over 20 years ago. The ConfigurationBuilder feature we introduced in .Net 4.7.1 had to stay within the bounds of this old and rigid config system. To be as flexible as possible, there were two points while constructing config sections that looked useful for extension. So ConfigurationBuilders get a crack at the raw xml when it is first read, and then again at the ConfigurationSection object that comes back from the factory that built it from the raw xml. So on any given section, all config builders tagged on that section get to process the xml string first, and then all config builders get to process a ConfigurationSection object.

If you are processing a config section other than appSettings, you don't have any limitation. Config builders are instantiated per-section, so while your AzureAppConfiguration builder is being created to process your WCF section, it goes to ConfigurationManager.AppSettings to look for its connection information. The config system does the whole song and dance on that section, instantiating the necessary config builders there, and produces a collection of application settings for your WCF-processing AzureAppConfiguration builder to use when initializing.

If you are processing appSettings with an Expand-mode builder... you're out of luck. When builders initialize to process appSettings, they can't call out to the config system to get the appSettings section. That just creates an infinitely recursive loop.

When running in Strict or Greedy mode, we can fake this a little bit by directly using the AppSettingSection object that is being processed instead of going back to the config system and asking for it. It's a little hacky, but it closes the gap for 99.9% of cases because really there shouldn't be any need to apply an Expand-mode builder to a section that is so simply key/value oriented.
